File: aws-globalaccelerator-accelerator/src/test/java/software/amazon/globalaccelerator/accelerator/CreateHandlerTest.java
package software.amazon.globalaccelerator.accelerator;


import com.amazonaws.services.globalaccelerator.model.*;
import lombok.val;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.ArrayList;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
public class CreateHandlerTest {

    @Mock
    private AmazonWebServicesClientProxy proxy;

    @Mock
    private Logger logger;

    @BeforeEach
    public void setup() {
        proxy = mock(AmazonWebServicesClientProxy.class);
        logger = mock(Logger.class);
    }

    @Test
    public void handleRequest_InitialStateCreatesAccelerator() {
        val result = new CreateAcceleratorResult()
                .withAccelerator(new Accelerator()
                    .withStatus(AcceleratorStatus.IN_PROGRESS)
                    .withEnabled(true)
                    .withAcceleratorArn("ACCELERATOR_ARN"));

        doReturn(result).when(proxy).injectCredentialsAndInvoke(any(CreateAcceleratorRequest.class), any());

        // create the input
        val handler = new CreateHandler();
        val tags = new ArrayList<Tag>();
        tags.add(software.amazon.globalaccelerator.accelerator.Tag.builder().key("K1").value("V1").build());
        val model = ResourceModel.builder().enabled(true).name("AcceleratorTest").tags(tags).build();
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();

        val response = handler.handleRequest(proxy, request, null, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getResourceModel()).isNotNull();
        assertThat(response.getResourceModel().getAcceleratorArn()).isEqualTo("ACCELERATOR_ARN");
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();
    }

    @Test
    public void handleRequest_PendingDeployedReturnsInProgress() {
        val result = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS)
                        .withEnabled(true)
                        .withAcceleratorArn("ACCELERATOR_ARN"));

        doReturn(result).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // create the input
        val handler = new CreateHandler();
        val tags = new ArrayList<Tag>();
        tags.add(software.amazon.globalaccelerator.accelerator.Tag.builder().key("K1").value("V1").build());
        val model = ResourceModel.builder()
                .enabled(true)
                .name("AcceleratorTest")
                .tags(tags)
                .acceleratorArn("ACCELERATOR_ARN").build();
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();

        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(10).build();
        val response = handler.handleRequest(proxy, request, callbackContext, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.IN_PROGRESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(1);
        assertThat(response.getCallbackContext()).isNotNull();
        assertThat(response.getCallbackContext().getStabilizationRetriesRemaining()).isEqualTo(callbackContext.getStabilizationRetriesRemaining()-1);
        assertThat(response.getResourceModel()).isNotNull();
        assertThat(response.getResourceModel().getAcceleratorArn()).isEqualTo("ACCELERATOR_ARN");
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();
    }

    @Test
    public void handleRequest_DeployedReturnsSuccess() {
        val result = new DescribeAcceleratorResult()
                .withAccelerator(new Accelerator()
                        .withStatus(AcceleratorStatus.DEPLOYED)
                        .withEnabled(true)
                        .withAcceleratorArn("ACCELERATOR_ARN"));

        doReturn(result).when(proxy).injectCredentialsAndInvoke(any(DescribeAcceleratorRequest.class), any());

        // create the input
        val handler = new CreateHandler();
        val tags = new ArrayList<Tag>();
        tags.add(software.amazon.globalaccelerator.accelerator.Tag.builder().key("K1").value("V1").build());
        val model = ResourceModel.builder()
                .enabled(true)
                .name("AcceleratorTest")
                .tags(tags)
                .acceleratorArn("ACCELERATOR_ARN").build();
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();

        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(10).build();
        val response = handler.handleRequest(proxy, request, callbackContext, logger);

        assertThat(response).isNotNull();
        assertThat(response.getStatus()).isEqualTo(OperationStatus.SUCCESS);
        assertThat(response.getCallbackDelaySeconds()).isEqualTo(0);
        assertThat(response.getCallbackContext()).isNull();
        assertThat(response.getResourceModel()).isNotNull();
        assertThat(response.getResourceModel().getAcceleratorArn()).isEqualTo("ACCELERATOR_ARN");
        assertThat(response.getResourceModels()).isNull();
        assertThat(response.getMessage()).isNull();
    }

    @Test
    public void handleRequest_StabilizationTooLongFails() {
        // create the input
        val handler = new CreateHandler();
        val model = ResourceModel.builder()
                .acceleratorArn("ACCELERATOR_ARN").build();
        val request = ResourceHandlerRequest.<ResourceModel>builder().desiredResourceState(model).build();

        val callbackContext = CallbackContext.builder().stabilizationRetriesRemaining(0).build();
        assertThatExceptionOfType(RuntimeException.class).isThrownBy(() -> handler.handleRequest(proxy, request, callbackContext, logger))
            .withMessageMatching("Timed out waiting for global accelerator to be deployed.");
    }
}

File: aws-globalaccelerator-accelerator/src/test/kotlin/software/amazon/globalaccelerator/accelerator/CreateHandlerTest.kt
package software.amazon.globalaccelerator.accelerator

import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus
import com.amazonaws.services.globalaccelerator.model.CreateAcceleratorResult
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorResult
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import java.util.*

@ExtendWith(MockKExtension::class)
class CreateHandlerTest {
    @MockK
    lateinit var proxy: AmazonWebServicesClientProxy

    @MockK(relaxed = true)
    lateinit var logger: Logger

    @BeforeEach
    fun setup() = MockKAnnotations.init(this)

    @Test
    fun handleRequest_InitialStateCreatesAccelerator() {
        val result = CreateAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS)
                        .withEnabled(true)
                        .withAcceleratorArn("ACCELERATOR_ARN"))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyCreateAccelerator>()) } returns result

        // create the input
        val handler = CreateHandler()
        val tags = listOf(Tag.builder().key("K1").value("V1").build())
        val model = ResourceModel.builder().enabled(true).name("AcceleratorTest").tags(tags).build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val response = handler.handleRequest(proxy, request, null, logger)
        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertNotNull(response.resourceModel)
        assertEquals("ACCELERATOR_ARN", response.resourceModel.acceleratorArn)
        assertNull(response.resourceModels)
        assertNull(response.message)
    }

    @Test
    fun handleRequest_PendingDeployedReturnsInProgress() {
        val result = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.IN_PROGRESS)
                        .withEnabled(true)
                        .withAcceleratorArn("ACCELERATOR_ARN"))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns result

        // create the input
        val handler = CreateHandler()
        val tags = listOf(Tag.builder().key("K1").value("V1").build())
        val model = ResourceModel.builder()
                .enabled(true)
                .name("AcceleratorTest")
                .tags(tags)
                .acceleratorArn("ACCELERATOR_ARN").build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(10)
        val response = handler.handleRequest(proxy, request, callbackContext, logger)
        assertNotNull(response)
        assertEquals(OperationStatus.IN_PROGRESS, response.status)
        assertEquals(1, response.callbackDelaySeconds)
        assertNotNull(response.callbackContext)
        assertEquals(response.callbackContext!!.stabilizationRetriesRemaining, callbackContext.stabilizationRetriesRemaining - 1)
        assertNotNull(response.resourceModel)
        assertEquals("ACCELERATOR_ARN", response.resourceModel.acceleratorArn)
        assertNull(response.resourceModels)
        assertNull(response.message)
    }

    @Test
    fun handleRequest_DeployedReturnsSuccess() {
        val result = DescribeAcceleratorResult()
                .withAccelerator(Accelerator()
                        .withStatus(AcceleratorStatus.DEPLOYED)
                        .withEnabled(true)
                        .withAcceleratorArn("ACCELERATOR_ARN"))
        every { proxy.injectCredentialsAndInvoke(ofType(), ofType<ProxyDescribeAccelerator>()) } returns result

        // create the input
        val handler = CreateHandler()
        val tags = listOf(Tag.builder().key("K1").value("V1").build())
        val model = ResourceModel.builder()
                .enabled(true)
                .name("AcceleratorTest")
                .tags(tags)
                .acceleratorArn("ACCELERATOR_ARN").build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(10)
        val response = handler.handleRequest(proxy, request, callbackContext, logger)
        assertNotNull(response)
        assertEquals(OperationStatus.SUCCESS, response.status)
        assertEquals(0, response.callbackDelaySeconds)
        assertNotNull(response.resourceModel)
        assertEquals("ACCELERATOR_ARN", response.resourceModel.acceleratorArn)
        assertNull(response.resourceModels)
        assertNull(response.message)
    }

    @Test
    fun handleRequest_StabilizationTooLongFails() {
        // create the input
        val handler = CreateHandler()
        val model = ResourceModel.builder()
                .acceleratorArn("ACCELERATOR_ARN").build()
        val request = ResourceHandlerRequest.builder<ResourceModel>().desiredResourceState(model).build()
        val callbackContext = CallbackContext(0)
        val exception = assertThrows(RuntimeException::class.java) {
            handler.handleRequest(proxy, request, callbackContext, logger)
        }
        assertEquals("Timed out waiting for global accelerator to be deployed.", exception.message)
    }
}
